### PR TITLE
[FLINK-25608][metrics] Annotate metrics classes with Public(Evolving)

### PR DESCRIPTION
**Pull request by fapaul**
## What is the purpose of the change

Our metrics classes are already treated as if they are public. This PR annotates the bases classes with `@Public` and the metrics helper exposed to the connectors with `@PublicEvolving`. We also fix some of our arch unit test by introducing the annotations

The change was initially required by #18302.

## Brief change log

- Annotate metrics classes with public annotations

## Verifying this change

- Arch unit tests still pass

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by fapaul** [comment]
@zentol I extracted the metrics API annotations to a separate ticket please have another look.

**Review by zentol — CHANGES_REQUESTED** [review]
How did you decide which classes should be public? I see that `Metric` and all extending interfaces are `Public`, yet `View` isn't despite being an extension for metrics.

Why are the reporter interfaces not annotated?

**Comment by fapaul** [comment]
> How did you decide which classes should be public? I see that `Metric` and all extending interfaces are `Public`, yet `View` isn't despite being an extension for metrics.
> 
> Why are the reporter interfaces not annotated?

Good question, I mainly went through the metrics classes relevant for connectors. Do you think all classes inside the `metrics-core` module should be annotated?

**Comment by zentol** [comment]
> Do you think all classes inside the metrics-core module should be annotated?

All interfaces + the `AbstractReporter/HistogramStatistics/MetricConfig` classes I think.
The remaining classes (SimpleCounter, MeterView, UnregisteredMetricsGroup) should be `@Internal`.

File: flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/LogicalScopeProvider.java [hunk 1]
```diff
@@ -17,13 +17,16 @@
...
+import org.apache.flink.annotation.Public;
+
 /**
  * Extension for metric groups that support logical scopes.
  *
  * <p>ATTENTION: This interface is *not* meant for the long-term; it merely removes the need for
  * reporters to depend on flink-runtime in order to access the logical scope. Once the logical scope
  * is properly exposed this interface *will* be removed.
  */
+@Public
```
Comment (fapaul) [inline flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/LogicalScopeProvider.java:29]: @zentol can you confirm this annotation?

**Review by fapaul — COMMENTED** [review]
I went through all the metrics classes and added the proper annotation. Thanks for your help.

File: flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/LogicalScopeProvider.java (hunk 1 above)
Comment (zentol) [reply flink-metrics/flink-metrics-core/src/main/java/org/apache/flink/metrics/LogicalScopeProvider.java:29]: yes this should be fine. While we do intend to drop it in 2.0, until then it should remain stable.

**Comment by zentol** [comment]
@flinkbot run azure

**Pull request merged**